sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcTableOperations.scala

```diff
@@ -0,0 +1,116 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.hive.orc
+
+import java.util._
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.Path
+import org.apache.hadoop.hive.ql.io.orc._
+import org.apache.hadoop.io.NullWritable
+import org.apache.hadoop.io.Writable
+import org.apache.hadoop.mapred.JobConf
+import org.apache.hadoop.mapreduce.Job
+import org.apache.hadoop.mapreduce.lib.input.FileInputFormat
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.hive.HiveShim
+import org.apache.spark.sql.sources.Filter
+import org.apache.spark.{Logging, SerializableWritable}
+import scala.collection.JavaConversions._
+
+case class OrcTableScan(attributes: Seq[Attribute],
+    @transient relation: OrcRelation,
+    filters: Array[Filter],
+    inputPaths: Array[String]) extends Logging {
+  @transient val sqlContext = relation.sqlContext
+  val path = relation.paths(0)
+
+  def addColumnIds(output: Seq[Attribute],
+                   relation: OrcRelation, conf: Configuration) {
+    val ids =
+      output.map(a =>
+        relation.dataSchema.toAttributes.indexWhere(_.name == a.name): Integer)
+        .filter(_ >= 0)
+    val names = attributes.map(_.name)
+    val sorted = ids.zip(names).sorted
+    HiveShim.appendReadColumns(conf, sorted.map(_._1), sorted.map(_._2))
+  }
```

Comment:
Actually how about

```scala
    val ids = output.map(a => relation.dataSchema.fieldIndex(a.name): Integer)
    val (sortedIds, sortedNames) = ids.zip(attributes.map(_.name)).sorted.unzip
    HiveShim.appendReadColumns(conf, sortedIds, sortedNames)
```

Especially, if a column name is not found in `dataSchema`, instead of `filter(_ >= 0)`, we should report an error, which is what `StructType.fieldIndex` does.